lightning/src/ln/onchain_utils.rs

```diff
@@ -1,8 +1,698 @@
+		where L::Target: Logger,
+	{
+		match self.malleability {
+			PackageMalleability::Malleable => {
+				let mut bumped_tx = Transaction {
+					version: 2,
+					lock_time: 0,
+					input: vec![],
+					output: vec![TxOut {
+						script_pubkey: destination_script,
+						value,
+					}],
+				};
+				for (outpoint, _) in self.inputs.iter() {
+					bumped_tx.input.push(TxIn {
+						previous_output: *outpoint,
+						script_sig: Script::new(),
+						sequence: 0xfffffffd,
+						witness: Vec::new(),
+					});
+				}
+				for (i, (outpoint, out)) in self.inputs.iter().enumerate() {
+					log_trace!(logger, "Adding claiming input for outpoint {}:{}", outpoint.txid, outpoint.vout);
+					if !out.finalize_input(&mut bumped_tx, i, onchain_handler) { return None; }
+				}
+				log_trace!(logger, "Finalized transaction {} ready to broadcast", bumped_tx.txid());
+				return Some(bumped_tx);
+			},
+			PackageMalleability::Untractable => {
+				if let Some((outpoint, outp)) = self.inputs.first() {
+					if let Some(final_tx) = outp.get_finalized_tx(outpoint, onchain_handler) {
+						log_trace!(logger, "Adding claiming input for outpoint {}:{}", outpoint.txid, outpoint.vout);
+						log_trace!(logger, "Finalized transaction {} ready to broadcast", final_tx.txid());
+						return Some(final_tx);
+					}
+					return None;
+				} else { panic!("API Error: Package must not be inputs empty"); }
+			},
+		}
+	}
+	pub (crate) fn build_package(txid: Txid, vout: u32, input_solving_data: PackageSolvingData, absolute_timelock: u32, aggregation: bool, feerate_previous: u64, height_timer: Option<u32>, height_original: u32) -> Self {
+		let malleability = match input_solving_data {
+			PackageSolvingData::RevokedOutput(..) => { PackageMalleability::Malleable },
+			PackageSolvingData::RevokedHTLCOutput(..) => { PackageMalleability::Malleable },
+			PackageSolvingData::CounterpartyHTLCOutput(..) => { PackageMalleability::Malleable },
+			PackageSolvingData::HolderHTLCOutput(..) => { PackageMalleability::Untractable },
+			PackageSolvingData::HolderFundingOutput(..) => { PackageMalleability::Untractable },
+		};
+		let mut inputs = Vec::with_capacity(1);
+		inputs.push((BitcoinOutPoint { txid, vout }, input_solving_data));
+		PackageTemplate {
+			inputs,
+			malleability,
+			absolute_timelock,
+			aggregation,
+			feerate_previous,
+			height_timer,
+			height_original,
+		}
+	}
+}
+
+impl Writeable for PackageTemplate {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		writer.write_all(&byte_utils::be64_to_array(self.inputs.len() as u64))?;
+		for (ref outpoint, ref rev_outp) in self.inputs.iter() {
+			outpoint.write(writer)?;
+			rev_outp.write(writer)?;
+		}
+		self.malleability.write(writer)?;
+		self.absolute_timelock.write(writer)?;
+		self.aggregation.write(writer)?;
```

Comment:
Is there a need to serialize `malleability` and `aggregation`? Can't they be inferred from the other data? Possibly other fields, too.

Reply:
Yep added new CounterpartyHTLCOfferedOutput/CounterpartyHTLCReceivedOutput to remove ambiguity around aggregation and thus enable inference at deserialization.